File: app/src/main/java/com/ladsoft/bakingapp/mvp/presenter/RecipePresenter.java
package com.ladsoft.bakingapp.mvp.presenter;

import com.ladsoft.bakingapp.entity.Recipe;
import com.ladsoft.bakingapp.mvp.RecipeMvp;

public class RecipePresenter extends Presenter<RecipeMvp.View> implements RecipeMvp.Model.Callback {
    private final RecipeMvp.Model model;

    private boolean activityCreation = true;
    private RecipeMvp.StateContainer state;

    public RecipePresenter(RecipeMvp.Model model) {
        this.model = model;
        model.setListener(this);
    }

    public void setData(RecipeMvp.StateContainer state) {
        this.state = state;
    }

    public RecipeMvp.StateContainer getData() {
        return this.state;
    }

    public void loadData() {
        if (!isViewAttached()) { return; }

        if (activityCreation) {
            getView().showRefresh(true);
            model.loadRecipe(state.getRecipeId());
        }
    }

    @Override
    public void detachView() {
        getView().showRefresh(false);
        super.detachView();
    }

    @Override
    public void onDataLoaded(Recipe recipe) {
        if (!isViewAttached()) { return; }

        getView().onRecipeLoaded(recipe);
        getView().showRefresh(false);
        activityCreation = false;
    }

    @Override
    public void onDataLoadError() {
        if (!isViewAttached()) { return; }

        getView().onRecipeLoadError();
        getView().showRefresh(false);
    }
}

File: app/src/main/java/com/ladsoft/bakingapp/mvp/presenter/RecipePresenter.kt
package com.ladsoft.bakingapp.mvp.presenter

import com.ladsoft.bakingapp.entity.Recipe
import com.ladsoft.bakingapp.mvp.RecipeMvp

class RecipePresenter(val model: RecipeMvp.Model): Presenter<RecipeMvp.View>(), RecipeMvp.Model.Callback {
    var activityCreation = true;

    init {
        model.setListener(this)
    }

    lateinit var state : RecipeMvp.StateContainer

    fun setData(state : RecipeMvp.StateContainer) {
        this.state = state
    }

    fun loadData() {
        when {
            activityCreation -> {
                view?.showRefresh(true)
                model.loadRecipe(state.recipeId)
            }
        }
    }

    override fun detachView() {
        view?.showRefresh(false)
        super.detachView()
    }

    override fun onDataLoaded(recipe: Recipe) {
        view?.onRecipeLoaded(recipe)
        view?.showRefresh(false)
        activityCreation = false
    }

    override fun onDataLoadError() {
        view?.onRecipeLoadError()
        view?.showRefresh(false)
    }
}